### PR TITLE
Initial checkin for using Function Executor to call local functions

### PR DESCRIPTION
**Pull request by omkarmore83**
These changes make use of the IFunctionExecutor functionality of proxies. Functions create an implementation of IFuncExecutor. This is passed to the IProxyClient. This will be used to call the local functions efficiently, instead of the current roundtrip.

### CODE REVIEW

**Comment by omkarmore83** [comment]
Working on EndToEnd tests to cover the scenario.

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            if (function == null)
+            {
+                // request does not map to an HTTP function
+                request.Properties[ScriptConstants.AzureFunctionsHttpResponseKey] = new HttpResponseMessage(HttpStatusCode.NotFound);
+                return;
+            }
+            request.SetProperty(ScriptConstants.AzureFunctionsHttpFunctionKey, function);
+
+            var authorizationLevel = await FunctionsController.DetermineAuthorizationLevelAsync(request, function, _dependencyResolver);
+            if (function.Metadata.IsExcluded ||
```
Comment (mamaso) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Can we remove the `IsExcluded` and `IsDisabled` checks here? I think these should already be handled in `TryParseFunctionMetadata` - the functions won't be added to the script host function collection.

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        private FunctionsController _functionsController;
+
+        internal ProxyFunctionExecutor(WebScriptHostManager scriptHostManager, IDependencyResolver dependencyResolver, FunctionsController functionsController)
+        {
+            _scriptHostManager = scriptHostManager;
+            _dependencyResolver = dependencyResolver;
+            _functionsController = functionsController;
+        }
+
+        public async Task ExecuteFuncAsync(string funcName, Dictionary<string, object> arguments, CancellationToken cancellationToken)
```
Comment (mamaso) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: This looks identical to `ExecuteAsync` in `FunctionsController` - can we refactor so there isn't that duplication?

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 3]
```diff
@@ -56,6 +58,9 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
             {
                 return await ProcessRequestAsync(req, function, ct);
             };
+
+            IFuncExecutor proxyFunctionExecutor = new ProxyFunctionExecutor(this._scriptHostManager, controllerContext.Configuration.DependencyResolver, this);
+            request.Properties.Add(ScriptConstants.AzureProxyFunctionExecutorKey, proxyFunctionExecutor);
             return await _scriptHostManager.HttpRequestManager.ProcessRequestAsync(request, processRequestHandler, cancellationToken);
```
Comment (mamaso) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: can we decide at this point whether we want to `ProcessRequestAsync` with the scriptHostManager or the proxyClient? I don't understand why attaching an executor to request properties is necessary.

Comment (omkarmore83) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: This may need more explanation. The change is trying to optimize the scenario where the backend of the proxy is another function. In such cases we want to avoid forwarding the request to the backend via frontend etc. Instead we pass a wrapper on FunctionsController (here ProxyFunctionExecutor) which will be used by the ProxyClient to reroute the request after it performs any request/response overrides. 

---
In reply to: [135931711](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r135931711) [](ancestors = 135931711)

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs [hunk 4]
```diff
@@ -418,7 +418,7 @@ internal static void AddRouteDataToRequest(IHttpRouteData routeData, HttpRequest
                     routeDataValues.Add(pair.Key, value);
                 }
 
-                request.Properties.Add(HttpExtensionConstants.AzureWebJobsHttpRouteDataKey, routeDataValues);
+                request.Properties[HttpExtensionConstants.AzureWebJobsHttpRouteDataKey] = routeDataValues;
```
Comment (MikeStall) [inline src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: Does this mean we're setting property multiple times now?

File: src/WebJobs.Script/Description/Proxies/ProxyClientExecutor.cs [hunk 5]
```diff
@@ -24,7 +24,12 @@ public ProxyData GetProxyData()
 
         public async Task Execute(HttpRequestMessage request, ILogger logger)
         {
-            await _proxyClient.CallAsync(new object[] { request }, null, logger);
+            IFuncExecutor proxyFunctionExecutor = null;
+            if (request.Properties.ContainsKey(ScriptConstants.AzureProxyFunctionExecutorKey))
```
Comment (MikeStall) [inline src/WebJobs.Script/Description/Proxies/ProxyClientExecutor.cs]: > if (request.Properties.ContainsKey(ScriptConstants.AzureProxyFunctionExecutorKey)) [](start = 11, length = 83)

nit: Could this be TryGetValue?  That makes it a single lookup call.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 6]
```diff
@@ -56,6 +58,9 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
             {
                 return await ProcessRequestAsync(req, function, ct);
             };
+
+            IFuncExecutor proxyFunctionExecutor = new ProxyFunctionExecutor(this._scriptHostManager, controllerContext.Configuration.DependencyResolver, this);
```
Comment (MikeStall) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: >this [](start = 153, length = 4)

I think passing this controller instance in here is incorrect. 
Say request #1 is the initial call to the proxy. That has a backend call to a custom function (and if I recall, may invoke multiple custom functions for a single request) 
So that could lead to subsequent request #2, #3 for Functions. 

Controllers are specific to a request.  This controller instance is for the incoming request (#1). But you're reusing it for request #2, #3 too. 
 
I think the correct think here is to split this controller out. Create a new non-controller class (ie, "FunctionRequestInvoker")  which has ProcessRequestAsync() below. 
Then both this controller and your IFuncExecutor can just calls that.

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 2 above)
Comment (MikeStall) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: >arguments [](start = 87, length = 9)

What's the semantics of this dictionary? What are the keys?

Comment (MikeStall) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: >funcName [](start = 50, length = 8)

Why isn't this parameter used?

Comment (omkarmore83) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: This is currently part of the contract definition. However, we cant really use this, as there is no way for the proxy code to determine which function would be called for a given request (e.g: the request could be using route definition, or have a route prefix). For this reason the proxy client sends the update request itself. and the ProxyFuncitonExecutor does the routing again.

---
In reply to: [136211157](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r136211157) [](ancestors = 136211157)

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 6 above)
Comment (omkarmore83) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: I was trying to minimize changes to the FunctionController. I can make a class FunctionRequestInvoker and move some common logic there as static.

---
In reply to: [136211006](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r136211006) [](ancestors = 136211006)

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs (hunk 4 above)
Comment (omkarmore83) [reply src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: Yes. If the call goes from proxies to functions. the request uri changes (as the backend changes) and this means that the routedata will need to be updated.

---
In reply to: [136209751](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r136209751) [](ancestors = 136209751)

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 2 above)
Comment (omkarmore83) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Currently the only key used is "AzureFunctionsHttpRequestKey". This allows us to pass the updated request to the FunctionExecutor.  

---
In reply to: [136211097](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r136211097) [](ancestors = 136211097)

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 1 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: IsExcluded can be removed. but not the check for IsDisabled because they can still be  called with admin level auth.

---
In reply to: [135930304](https://github.com/Azure/azure-webjobs-sdk-script/pull/1839#discussion_r135930304) [](ancestors = 135930304)

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs (hunk 4 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: You should add a new test for this since you rely on this new behavior and don't want it to regress. See the unit tests we already have in place for this method.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 7]
```diff
@@ -33,110 +35,26 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
...
-                    return new HttpResponseMessage(HttpStatusCode.Unauthorized);
-                }
-
-                // Not a WebHook request so dispatch directly
-                response = await _scriptHostManager.HandleRequestAsync(function, request, cancellationToken);
-            }
-
-            return response;
+            IFuncExecutor proxyFunctionExecutor = new ProxyFunctionExecutor(this._scriptHostManager, controllerContext.Configuration.DependencyResolver, _webHookReceiverManager);
+            request.Properties.Add(ScriptConstants.AzureProxyFunctionExecutorKey, proxyFunctionExecutor);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Why do you have to pass this as a request property? Seems like this should just be part of ProxyClientExecutor - why can't it be set up with this to begin with, rather than passing it to the executor this way?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                request.Properties[ScriptConstants.AzureFunctionsHttpResponseKey] = response;
+                return;
+            }
+
+            Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> processRequestHandler = async (req, ct) =>
+            {
+                return await FunctionRequestInvoker.ProcessRequestAsync(req, function, ct, _scriptHostManager, _webHookReceiverManager);
+            };
+
+            var resp = await _scriptHostManager.HttpRequestManager.ProcessRequestAsync(request, processRequestHandler, cancellationToken);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Should you be pumping this into the HttpRequestManager again here? If the request has already been initially processed by the FunctionsController, it has already been through the manager throttles - wouldn't this mean it would happen twice?

File: src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+        public static async Task<HttpResponseMessage> PreValidateRequestAsync(HttpRequestMessage request, FunctionDescriptor function, IDependencyResolver resolver)
+        {
+            if (function == null)
+            {
+                if (request.RequestUri.AbsolutePath == "/")
+                {
+                    // if the request is to the root and we can't find any matching FunctionDescriptors which might have been setup by proxies
+                    // then homepage logic will be applied.
+                    return (IsHomepageDisabled || request.IsAntaresInternalRequest())
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs]: The naming of this method isn't correct, since here we're doing more than validation - we're returning the home page

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 7 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: ProxyClientExecutor only gets created when there are proxies present and we are about to invoke a proxy. at this point we don't know if we the incoming url will be processed by proxy or not so just save the context in the request property.

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 8 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Yes and this is expected. the incoming request is for a proxy endpoint and in proxy the request properties might have been changed with requestoverrides so once all the overrides have been processed and we have determined this is a local function, we will need to pump it thru again. the main purpose of this PR is not to send the full request to FEs and such and hit sandbox limitations, ... when the call is to a local function.

File: src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs (hunk 9 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs]: Good point, I am going to rename it to PreProcessRequestAsync. Proxy is just returning what its route is configured to return. imagine a scenario where the route is to a local funaction and in request overrides it gets changed to rhe root of the function app. we just need to be consistent and make sure the same code path runs for both.

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs (hunk 4 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: Good point, I will update the existing test to make sure this property is always set when routeData.Values != null

File: test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTests.cs [hunk 10]
```diff
@@ -237,6 +237,8 @@ public void AddRouteDataToRequest_AddsRequestProperty_WhenRouteDataNotNull()
 
             WebScriptHostManager.AddRouteDataToRequest(mockRouteData.Object, request);
 
+            Assert.True(request.Properties.ContainsKey(HttpExtensionConstants.AzureWebJobsHttpRouteDataKey));
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTests.cs]: This additional check doesn't seem to add anything or test what you're trying to test - we already know that the value is not null, because we're accessing it below. What you want to test is that you can call AddRouteDataToRequest a second time and that the returned instance is updated, right?

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 7 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Ok. Is there a way to only create/add this executor if proxies are enabled or being used?

Comment (safihamid) [reply src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Yes we won't know what we route to is a local function or external url here but we do know if the incoming route is for a proxy or not. I will add a check for function.Metadata.IsProxy here when creating/adding this executor.

File: src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                }
+
+                // Not a WebHook request so dispatch directly
+                response = await scriptHostManager.HandleRequestAsync(function, request, cancellationToken);
+            }
+
+            return response;
+        }
+
+        public static async Task<HttpResponseMessage> PreprocessRequestAsync(HttpRequestMessage request, FunctionDescriptor function, IDependencyResolver resolver)
```
Comment (MikeStall) [inline src/WebJobs.Script.WebHost/FunctionRequestInvoker.cs]: >IDependencyResolver [](start = 134, length = 19)

Passing the depednency resolver around seems suspect ... can we  pass just the interfaces we need (ISecretManager?) .
This is a static method; could this be an instance method and then the require interfaces are properties?
